src/identifiers/logic.py

```diff
@@ -120,28 +133,41 @@ def get_dois_for_articles(articles, create=False):
     return identifiers
 
 
-def poll_dois_for_articles(articles):
+def poll_dois_for_articles(articles, **kwargs):
     clear_cache()
+
+    start = kwargs.pop('start', None)
+    if not start:
+        start = datetime.datetime.now()
+
     status = ''
     error = False
     identifiers = get_dois_for_articles(articles)
     polled = set()
-    for identifier in identifiers:
+    for i, identifier in enumerate(identifiers):
+
+        # Time out gracefully
+        if datetime.datetime.now() > start + datetime.timedelta(seconds=30):
```

Comment:
It seems to me that the 30 seconds used here are matching the cache invalidation term used on the decorators above.
If so, it would be more explicit to have a constant at the top of this module like:

`CROSSREF_TIMEOUT_SECONDS = 30`

Also, we want the timeout to be an optional argument to this function, since it is only relevant when executing under an HTTP request, but not when executing from command line or a Django command. (The default can be `timeout=CROSSREF_TIMEOUT_SECONDS`) which would allow command line users to set it as `timeout=None`

Finally, you don't need to work with datetime objects here, it is best to work with `time.time()` (integers) since you are replicating the behaviour of a timer here. e.g.:

```
start_time = time.time()
for item in tems:
    #do some work
    if time.time() - start_time > 30:
         break
```
Or you could even write a `Timer` context manager for bonus pythonic points ;)